### PR TITLE
Extend `plot` method in LRFinder for multiple `param_groups`

### PR DESCRIPTION
**Pull request by KickItLikeShika**
`plot` had issues handling multiple `suggested_lrs`, this PR fixes this issue, and added `display_legends` arg.

Some explanations for the updates:
- Added `display_legends` to make it more clear to know which `suggested_lr` belongs to which `param_group`, and make it optional for the users to enable or disable it, in more complex cases maybe a lot legends would affect the plot it would make less clear.
- Ignored `plt.xlim()` in case of multiple `param_groups`, as we see in this picture it may we should keep as it is will make it more cleaner, I tried this:
```python
if isinstance(lrs[0], list):
     xlim1 = lrs[0].index(min(lrs[0]))
     xlim2 = lrs[-1].index(max(lrs[-1]))
     plt.xlim([lrs[0][xlim1], lrs[-1][xlim2]])
else:
     plt.xlim([lrs[0], lrs[-1]])
```
But I thought no need to add additional check as the result will be almost the same, as the other curves may start from another point.
- In `test_plot_multiple_param_groups()` I didn't make the model or anything else `pytest.fixtures()`, as we won't use them again, no need for that.

adsfasdfa

For more details about the experimentation: https://colab.research.google.com/drive/1xJfh8kMe3h7Ceyo9B4W4QSbxp-jUJwFT?usp=sharing

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by KickItLikeShika** [comment]
The reason behind this failure I think is the `y` tensor is all zeros https://github.com/pytorch/ignite/pull/1981/checks?check_run_id=2480178422#step:12:843 , increasing the sizes will solve these issues, I can solve it in this PR..

**Review by vfdev-5 — COMMENTED** [review]
@KickItLikeShika thanks for the PR ! I left few comments

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 1]
```diff
@@ -295,6 +296,58 @@ def test_plot(lr_finder, to_save, dummy_engine, dataloader):
...
+
+    with lr_finder.attach(trainer, to_save) as trainer_with_finder:
+        trainer_with_finder.run(dataloader)
+
+    # Avoid RuntimeError
+    if torch.tensor(lr_finder._history["loss"]).argmin() < 2:
+        lr_finder._history["loss"].insert(0, 10)
+        lr_finder._history["loss"].insert(0, 100)
+        lr_finder._history["lr"].insert(0, [10, 10, 10])
+        lr_finder._history["lr"].insert(0, [100, 100, 100])
```
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: Sometimes we raise this error
```
if len(decreasing_losses) == 1:
            raise RuntimeError("FastaiLRFinder got unexpected curve shape, the curve should be somehow U-shaped")
```
We avoid this RuntimeError by adding some large values at the beginning, and if you're wondering about the shape of these values, in case of multiple `param_groups`, if `len(param_groups) = 3` then `loss = [4, 3, 2, 1]` and `lrs = [[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]]`

File: ignite/contrib/handlers/lr_finder.py [hunk 2]
```diff
@@ -249,11 +253,20 @@ def plot(
...
+                for lr in sug_lr:
+                    ax.scatter(
+                        lr, corresponding_loss, s=75, marker="o", zorder=3,
+                    )
+                # Edit legends
+                legends = ["suggested lr for param_groups " + str(i) for i in range(len(sug_lr))]
+            else:
+                ax.scatter(
+                    sug_lr, corresponding_loss, s=75, marker="o", color="red", zorder=3,
+                )
```
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply ignite/contrib/handlers/lr_finder.py:269]: Such a great idea, Thanks!

File: ignite/contrib/handlers/lr_finder.py [hunk 3]
```diff
@@ -263,11 +276,15 @@ def plot(
...
+        # Check to display legends
+        if display_legends:
+            plt.legend(legends)
         # Plot loss as a function of the learning rate
         plt.plot(lrs, losses)
         if log_lr:
             plt.xscale("log")
-        plt.xlim([lrs[0], lrs[-1]])
+        if not isinstance(lrs[0], list):
+            plt.xlim([lrs[0], lrs[-1]])
```
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply ignite/contrib/handlers/lr_finder.py:287]: Yes we can, but It may look something like this
dummy_muliple_param_groups
That's why I thought we don't need it, What do you think?

File: tests/ignite/contrib/handlers/test_lr_finder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: And regarding the tests, please tell me what do you think about the model and optimizer for `test_plot_multiple_param_groups`, should we move them and make them `pytest.fixture()`? I don't know if we are going to use them in some other tests or not, I think in the next PR `apply_suggested_lr` It will be good to add a test for multiple `param_groups`

File: ignite/contrib/handlers/lr_finder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/contrib/handlers/lr_finder.py:287]: This looks a bit strange as blue, orange curves are also cut even if we cut only LR...

File: tests/ignite/contrib/handlers/test_lr_finder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: OK, 
1) the runtime error should be checked in a separate test.
2) let's set up everything such that we do not have runtime error and we do not insert any data a posteriori.

Lines 327-332 looks very strange, no helpful comment explaining why we are doing below lines. This is very bad practice to code like that. Please, avoid that.


> And regarding the tests, please tell me what do you think about the model and optimizer for test_plot_multiple_param_groups, should we move them and make them pytest.fixture()? I don't know if we are going to use them in some other tests or not, I think in the next PR apply_suggested_lr It will be good to add a test for multiple param_groups

Yes, we can factorize them. Good idea !

File: ignite/contrib/handlers/lr_finder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply ignite/contrib/handlers/lr_finder.py:287]: I think it's not strange, as we cut from the minimum LR to the max one

File: tests/ignite/contrib/handlers/test_lr_finder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: - We already check the runtime error in this test  https://github.com/pytorch/ignite/blob/91953b11a4c44909f6497b6c60c85ee6f3121d41/tests/ignite/contrib/handlers/test_lr_finder.py#L249 . 
- Any idea to reshape the data in that way?
- Fine, I will start working on that refactoring them..

File: ignite/contrib/handlers/lr_finder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/contrib/handlers/lr_finder.py:287]: OK, I think I understand now why it is so. I think it is OK to have the plot like that as the docs on skip_start and skip_end explicitly says that we skip batches. 

> Yes we can, but It may look something like this

OK, let's display it like that.

File: tests/ignite/contrib/handlers/test_lr_finder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: > So we better find an idea to save us the inserting part?

If we do not have sufficient lr/loss history, just add more iters to run. Maybe, we can also make the model smaller by reducing the number of hidden layers. I let you explore that.

[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: @vfdev-5 Okay I will explore it more, but I don't think if there is any other way to prevent the early points of the loss history to be the minimum, what do you think?

[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: I think we can make the NN diverge also by choosing larger max LRs.

[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: Okay, I will look into that, Thanks for the ideas!

[the remark this answers is not in the document]
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: I have been trying to find out a way to avoid this RuntimeError in `test_plot_single_param_group` first, the model already has one layer so we can't reduce it anymore, I tried to manipulate the learning rate with different values, but this didn't work too. I think manipulating the loss history (with the current way or some other way) is the most solid solution

[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_lr_finder.py:332]: > I think manipulating the loss history (with the current way or some other way) is the most solid solution

I would not call it "solid solution". Either you create a synthetic history from the begining or either you work with a model and get real loss values but do not mix both.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @KickItLikeShika !

In a follow-up PR let's extend linear mode LR scheduling
